File: sponge-jython/examples/script/py/correlators_builder.py
"""
Sponge Knowledge base
Correlator builders
"""

from java.util.concurrent.atomic import AtomicInteger
from java.util import Collections, LinkedHashMap

def onInit():
    # Variables for assertions only
    sponge.setVariable("hardwareFailureScriptCount", AtomicInteger(0))
    sponge.setVariable("hardwareFailureScriptFinishCount", AtomicInteger(0))
    sponge.setVariable("eventLogs", Collections.synchronizedMap(LinkedHashMap()))

def onLoad():
    def onEvent(correlator, event):
        eventLog = sponge.getVariable("eventLogs").get(correlator.meta.name)

        eventLog.append(event)
        correlator.logger.debug("{} - event: {}, log: {}", correlator.hashCode(), event.name, str(eventLog))
        sponge.getVariable("hardwareFailureScriptCount").incrementAndGet()
        if len(eventLog) == 4:
            sponge.getVariable("hardwareFailureScriptFinishCount").incrementAndGet()
            correlator.finish()

    sponge.enable(CorrelatorBuilder("SampleCorrelator").withEvents(["filesystemFailure", "diskFailure"]).withMaxInstances(1)
                .withOnAcceptAsFirst(lambda  correlator, event: event.name == "filesystemFailure")
                .withOnInit(lambda  correlator: sponge.getVariable("eventLogs").put(correlator.meta.name, []))
                .withOnEvent(onEvent))

def onStartup():
    sponge.event("filesystemFailure").set("source", "server1").send()
    sponge.event("diskFailure").set("source", "server1").send()
    sponge.event("diskFailure").set("source", "server2").send()
    sponge.event("diskFailure").set("source", "server1").send()
    sponge.event("diskFailure").set("source", "server2").send()

File: sponge-nashorn/examples/script/js/correlators_builder.js
/**
 * Sponge Knowledge Base
 * Correlator builders
 */

var AtomicInteger = java.util.concurrent.atomic.AtomicInteger
var Collections = java.util.Collections
var LinkedHashMap = java.util.LinkedHashMap

function onInit() {
    // Variables for assertions only
    sponge.setVariable("hardwareFailureScriptCount", new AtomicInteger(0))
    sponge.setVariable("hardwareFailureScriptFinishCount", new AtomicInteger(0))
    sponge.setVariable("eventLogs", Collections.synchronizedMap(new LinkedHashMap()))
}

function onLoad() {
    sponge.enable(new CorrelatorBuilder("SampleCorrelator").withEvents(["filesystemFailure", "diskFailure"]).withMaxInstances(1)
            .withOnAcceptAsFirst(function(correlator, event) { return event.name == "filesystemFailure"})
            .withOnInit(function(correlator) { sponge.getVariable("eventLogs").put(correlator.meta.name, []) })
            .withOnEvent(function(correlator, event) {
                eventLog = sponge.getVariable("eventLogs").get(correlator.meta.name)

                eventLog.push(event);
                correlator.logger.debug("{} - event: {}, log: {}", correlator.hashCode(), event.name, eventLog.toString());
                sponge.getVariable("hardwareFailureScriptCount").incrementAndGet();
                if (eventLog.length >= 4) {
                    sponge.getVariable("hardwareFailureScriptFinishCount").incrementAndGet();
                    correlator.finish();
                }
            }))
}

function onStartup() {
    sponge.event("filesystemFailure").set("source", "server1").send();
    sponge.event("diskFailure").set("source", "server1").send();
    sponge.event("diskFailure").set("source", "server2").send();
    sponge.event("diskFailure").set("source", "server1").send();
    sponge.event("diskFailure").set("source", "server2").send();
}